### ISSUE TITLE
[BUG] CIPHERTEXT not including timestamp, CIPHERTEXT messages not being emitted for all messages

### ISSUE THREAD

**Issue by burstfreeze** [adiwajshing/Baileys#919]
I was able to check https://github.com/adiwajshing/Baileys/commit/d2e3abfe49088777dcd20256be7f282868957638 but found some messages to still carry the timestamp of connection, and other messages being at a timestamp of `0`.

I went through my backend and noticed I set 0 as a placeholder if no timestamp is available. 

So then I went to check the Baileys log and there are 2 situations arising:

1) Baileys emits a `CIPHERTEXT` message prior to the actual message but it won't include a timestamp in that (see below)
2) Baileys does not emit a `CIPHERTEXT` message at all (hence no timestamp to see in first place)

1 looks like 

```{"messages":[{"key":{"remoteJid":"","fromMe":false,"id":"","participant":""},"messageStubType":"CIPHERTEXT","messageStubParameters":["Error: No sender key for: "]}],"type":"append"}```

2 simply emits the normal message without being preceded by a timestamp message

Your implementation does not look much different from my PR regarding the attributes but for some reason with the Baileys commit I do not receive the expected data. Have you had a chance to verify whether it works as expected for you? Asking because for me it did not emit a single non-zero timestamp (or in some cases no CIPHERTEXT message whatsoever)

**Comment by adiwajshing** [comment]
@burstfreeze you might want to enable "trace" (`logger.level = 'trace'`) logging to check what the original message stanza that failed looks like.

Can then check where the parsing is failing

**Comment by burstfreeze** [comment]
@adiwajshing most of the time I am on trace for you since I don't want to miss anything when submitting reports 😅

The messages in question did not occur anywhere else in the logging except for the actual message (no previous ciphertext with that ID) or as shown the ciphertext without a timestamp at all. This was with a session that I kept offline for 3 days to verify.

Today with a session that I kept offline for one night for testing this with your latest commits (in case it could also fix something regarding this) and my logging, I did catch one retry and it actually emitted the correct timestamp in the ciphertext unlike before :) The question is what happens when I keep it offline for days again. **Do you know how to force this issue?̉** I am not sure if I understand the auth file correctly, could I force this behavior by modifying keys to a bad state (or simply not updating keys at all) in order to trigger this?

Is there a chance of a ciphertext message being emitted anywhere else except for the function that you have modified?

In any case, progress has been made, this looks very promising, thank you so much for your work so far - maybe if you can tell me a way to force this I can be more elaborate about the circumstances of this issue should it arise

**Comment by adiwajshing** [comment]
Thanks man

So there must be the underlying binary message that has the cipher text because the parsing needs to happen on something -- so can you please double check?

**Comment by burstfreeze** [comment]
@adiwajshing is there a possibility of a log for a message in a way that the message does not include the timestamp? What I did was to filter all log lines with the message ID in question. If the issue lays somewhere higher upstream I may have to look through the byte arrays I suppose?

**Comment by adiwajshing** [comment]
Looking for the message ID should work -- but the message must be there for sure

**Comment by frmdeveloper** [comment]
> I was able to check [d2e3abf](https://github.com/adiwajshing/Baileys/commit/d2e3abfe49088777dcd20256be7f282868957638) but found some messages to still carry the timestamp of connection, and other messages being at a timestamp of `0`.
> 
> I went through my backend and noticed I set 0 as a placeholder if no timestamp is available.
> 
> So then I went to check the Baileys log and there are 2 situations arising:
> 
> 1. Baileys emits a `CIPHERTEXT` message prior to the actual message but it won't include a timestamp in that (see below)
> 2. Baileys does not emit a `CIPHERTEXT` message at all (hence no timestamp to see in first place)
> 
> 1 looks like
> 
> `{"messages":[{"key":{"remoteJid":"","fromMe":false,"id":"","participant":""},"messageStubType":"CIPHERTEXT","messageStubParameters":["Error: No sender key for: "]}],"type":"append"}`
> 
> 2 simply emits the normal message without being preceded by a timestamp message
> 
> Your implementation does not look much different from my PR regarding the attributes but for some reason with the Baileys commit I do not receive the expected data. Have you had a chance to verify whether it works as expected for you? Asking because for me it did not emit a single non-zero timestamp (or in some cases no CIPHERTEXT message whatsoever)

```js
logger.level = 'silent'
```
Finished. Hehehe

**Comment by burstfreeze** [comment]
> Looking for the message ID should work -- but the message must be there for sure

@adiwajshing I have not noticed a single message without a timestamp since. Not saying there were none but what's possible is that the ciphertext message was sent before your update so the message that seemingly did not have a timestamp was simply "too old" to show up in my db.

Is there any way to know whether a message was re-sent so I won't have to check for a matching ciphertext message each time?

**Comment by adiwajshing** [comment]
Nah man, that's not possible. If you check the listener for the message (`CB:message`) -- the logic prevents what you're describing from happening

The CIPHERTEXT stubtype message unambiguously tells you that the message received failed to be decrypted

**Comment by burstfreeze** [comment]
Couldn't do much testing these days but I haven't noticed any more messages without a previous timestamp message so it's probably been older messages that were missed when using an older commit, **but all messages received with the fixed commit seemed to appear correctly so it seems to be fine now**. Closing this, awesome job

**Comment by burstfreeze** [comment]
I have to re-open, just had a bad message timestamp in my database 😅

I immediately went to check whether I have saved any ciphertext messages for that message ID, which I did and it turns out there was not just one but two of them!

One had the correct timestamp, the other one the timestamp of the connection. 

So for some reason Baileys has emitted two timestamp messages.

If I understand correctly, the messages carrying the correct timestamp for a message ID are the ciphertext messages. I do

`if (msg.messageStubType == proto.WebMessageInfo.WebMessageInfoStubType.CIPHERTEXT) {`


Can there be an edge case where there is another "kind" of message that also has that `messageStubType` but has another "shape"? In other words, should I check these messages for something else than CIPHERTEXT in order to be able to distinguish these messages from each other so I can rely on the "correct" one?
